### ISSUE TITLE
Consider moving to petit parser

### ISSUE THREAD

**Issue by lyncodev** [jtwig/jtwig#352]
https://github.com/petitparser/java-petitparser

This would allow Jtwig to drop it's dependency on ASM.

**Comment by ebuildy** [comment]
+1: cant use jtwig on a Apache Spark program, because Spark has dependency for ASM 3.1, hence the error

``java.lang.IncompatibleClassChangeError: class org.objectweb.asm.tree.ClassNode has interface org.objectweb.asm.ClassVisitor as super class``

Or

``Exception in thread "main" java.lang.NoSuchMethodError: org.objectweb.asm.tree.ClassNode.<init>(I)V``

**Comment by lyncodev** [comment]
Parsky will be the choice

**Comment by bohnman** [comment]
@lyncodev Would require rewriting your parser, but I'm a fan of ANTLR.  I've written a number of grammars with it and it works really well.  http://www.antlr.org

**Comment by joao-de-melo** [comment]
Ryan, it would be awesome to see a parser implementation of the core
library in ANTLR :D

**Comment by IARI** [comment]
Browsing the core repo, I couldn't locate the grammar file used to generate the parser for the core library. Could someone point me to it?

**Comment by bohnman** [comment]
@IARI There is no grammar file.  The grammar is implemented solely in code, originally based on the [parboiled library](https://github.com/sirthias/parboiled/wiki/Simple-Java-Example)

**Comment by IARI** [comment]
Oh ... Yeah, I've started messing with jtwigs parboiled parser, trying to implement my own little extension (implementing some AddonParser).
But that turned out to be hard and error prone with lots of boilderplate involved.

I cannot promise that I can pull it through, since i have no experience with antlr at all - but where would you recommend me to start, if i wanted to write a grammar file for it?

**Comment by bohnman** [comment]
@IARI The documentation and presentation is a good start, which you can find [here](https://github.com/antlr/antlr4/blob/master/doc/index.md).  I've also found it helpful to look at a grammar file for a language I know.  You can find a directory of examples [here](https://github.com/antlr/grammars-v4).  The json grammar is really simple.  Finally, I have a GitHub project that utilizes a custom grammar, which can be found [here](https://github.com/bohnman/squiggly-filter-jackson/blob/master/src/main/antlr4/com/github/bohnman/squiggly/parser/antlr4/SquigglyExpression.g4).  hope this helps!

**Comment by bohnman** [comment]
@IARI I should also note that IntelliJ has a good ANTLR plugin where you can test your grammar.

**Comment by cjbrooks12** [comment]
Just want to point out that a big feature of Twig's grammar being defined in code (both PHP and JTwig) is that new language features or DSLs can be added by 3rd party plugins with very little restriction on their syntax. I haven't worked enough with Antlr to know if it supports this kind of parsing (runtime-configurable grammars/custom token sequences), but it's something to keep in mind.

**Comment by bohnman** [comment]
@cjbrooks12 Great point.  I've only had experience with ANTLR in the context of static grammars so I cannot provide insight on this either.

**Comment by MaZderMind** [comment]
I started experimenting with Antlr for JTwig in https://github.com/MaZderMind/antlr-jtwig but had not much time for it. I am able to parse and evaluate complex expressions, much better than with the current jtwig parser (ie. something like foo[bar[moo.qoo * 2].too][0][1+bar].bar[foo] which is impossible with the current parser.

I was not yet able to find code islands in a docuemnt and switch lexer modes then, but I know that it is possible.

Using antlr would give us a lot of great tooling, from the built in left-recursion-removal to the built-in test-rig (https://github.com/mobileink/lab.clj.antlr/blob/master/doc/testrig.md), railroad-generators (https://github.com/bkiers/rrd-antlr4) and tons of examples (https://github.com/antlr/grammars-v4).

It's still a complex library and while a lot of things are very easy and require not much code at all, some are not so trivial and well documented (like switching lexer-mode from code-islands-in-document-mode to expression-mode). Nevertheless I would absolutely vote for Antlr, knowing that this would basicly mean a rewrite of 2/3 of JTwig.

**Comment by cjbrooks12** [comment]
@MaZderMind https://github.com/bkiers/Liqp uses Antlr for a Liquid implementation, whose syntax is very similar to Twig, and would probably be very helpful for learning how to do code islands and custom tags

**Comment by bohnman** [comment]
@MaZderMind @cjbrooks12 Agreed.  Just note that the grammar there is in Antlr version 3.  Version 4 is a little bit different.
